[tmva][sofie] New Keras Parser

PrasannaKasar opened the pull request:
# This Pull request:
This pull request adds the support for parsing Keras models with SOFIE

## Changes or fixes:
Currently, SOFIE's existing Keras parser is written in C++ and is quite old. Although it's written in C++ but its actual parsing logic is written in Python. Additionally, it lacks support for parsing layers such as MaxPool. 
This new parser is natively written in Python and uses of pythnoization to access C++ methods from SOFIE. The parser support latest version of Keras (Keras 3) and also has backwards compatibility with earlier versions of Keras 2. It also supports for both types of models i.e. models built using Keras' Functional as well as Sequential API.

Layers supported by the parser:
- [x] Add
- [x] AveragePool2D channels first
- [x] AveragePool2D channels last
- [x] BatchNorm
- [x] Concat
- [x] Conv2D channels first
- [x] Conv2D channels last
- [x] Conv2D padding same
- [x] Conv2D padding valid
- [ ] Conv2D transpose
- [x] Dense
- [ ] Dropout / Identity
- [x] Elu
- [x] Flatten
- [x] GlobalAveragePool2d channels first
- [x] GlobalAveragePool2d channels last
- [ ] GRU
- [x] LayerNorm
- [x] LeakyReLU
- [ ] LSTM
- [x] MaxPool2D channels first
- [x] MaxPool2D channels last
- [x] Multiply
- [x] Permute
- [x] Relu
- [x] Reshape
- [x] Selu
- [x] Sigmoid
- [ ] SimpleRNN
- [x] Softmax
- [x] Subtract
- [x] Swish
- [x] Tanh


## Checklist:

- [x] tested changes locally

sanjibansg commented:
> Thank you very much for the PR! It's indeed better to implement this in Python.
> 
> > Currently, SOFIE's existing Keras parser is written in C++ and is quite old.
> 
> While waiting for the review by @lmoneta and @sanjibansg, I have a general question.
> 
> What is the plan with the existing Keras parser in C++? If it is superseded by your new parser, can we remove it maybe in this PR as well? Like this we ensure that the users are not confused by two implementations, and also that the total maintenance cost doesn't increase.

I agree with @guitargeek, we will remove the existing C++ parser for Keras within this PR.

sanjibansg commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_sofie/_parser/_keras/layers/conv.py:
```diff
@@ -0,0 +1,70 @@
+        return op
+    else:
+        raise RuntimeError(
+            "TMVA::SOFIE - Unsupported - Operator Gemm does not yet support input type " + fLayerDType
```
```suggestion
            "TMVA::SOFIE - Unsupported - Operator Conv does not yet support input type " + fLayerDType
```

sanjibansg commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_sofie/_parser/_keras/layers/binary.py:
```diff
@@ -0,0 +1,23 @@
+          op =  gbl_namespace.TMVA.Experimental.SOFIE.ROperator_BasicBinary(float,'TMVA::Experimental::SOFIE::EBasicBinaryOperator::Mul')(fX1, fX2, fY)
+    else:
+        raise RuntimeError(
+            "TMVA::SOFIE - Unsupported - Operator Identity does not yet support input type " + fLayerDType
```
```suggestion
            "TMVA::SOFIE - Unsupported - Operator BasicBinary does not yet support input type " + fLayerDType
```

sanjibansg commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_sofie/_parser/_keras/layers/batchnorm.py:
```diff
@@ -0,0 +1,48 @@
+from cppyy import gbl as gbl_namespace
```
maybe good to check if the datatypes for the input tensors are float since we don't support the other cases?

sanjibansg commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_sofie/_parser/_keras/layers/concat.py:
```diff
@@ -0,0 +1,11 @@
+    input = [str(i) for i in finput]
+    output = str(foutput[0])
+    axis = int(attributes["axis"])
+    op =  gbl_namespace.TMVA.Experimental.SOFIE.ROperator_Concat(input, axis, 0,  output)
```
maybe good to check the datatype for the input tensor?

guitargeek reviewed (commented):
TensoFlow/Keras is too fragile to import unconditionally. You need to make sure it's not imported unconditionally, otherwise its presence will break several ROOT usecases, as we see in the tests. Always importing keras will also slow down importing ROOT, which is not desired

guitargeek commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_sofie/_parser/_keras/parser.py:
```diff
@@ -0,0 +1,479 @@
+from ......_pythonization import pythonization
+from cppyy import gbl as gbl_namespace
+import keras
```
Here for example: instead of importing `keras` whenever the parsers are imported, you need to import it locally in the functions.

PrasannaKasar replied:
hi @guitargeek, thanks for noticing this
but the reason I'm importing `keras` in this file is because I need the it to load the saved model using the input model file path

guitargeek commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/__init__.py:
```diff
@@ -44,6 +44,7 @@ def inject_rbatchgenerator(ns):
 
 
 from ._gnn import RModel_GNN, RModel_GraphIndependent
+from ._sofie._parser._keras.parser import RModelParser_Keras
```
But here is the main culprit: this global import is the entry point for all other imports I think.

PrasannaKasar replied:
just a question, if we don't expose `RModelParser_Keras`  class in the `_tmva` module, then how can we use it using `pythonization`?
I tried removing it but while using the parser it gave me this error
<img width="1323" height="132" alt="image" src="https://github.com/user-attachments/assets/ea643b0e-8c54-4793-b250-064fc93da344" />

pcanal commented on bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/__init__.py:
```diff
@@ -44,6 +44,7 @@ def inject_rbatchgenerator(ns):
 
 
 from ._gnn import RModel_GNN, RModel_GraphIndependent
+from ._sofie._parser._keras.parser import RModelParser_Keras
```
```suggestion
```

guitargeek commented:
> Also, @guitargeek, @pcanal can you please take a look at the import statements of keras and whether they are suitable with ROOT now?

I started the CI. If there are no seemingly unrelated test failures because of unconditionally importing keras anymore, then we know it's okay!

PrasannaKasar commented:
Hi @guitargeek 
Apologies I couldn't update you on this issue
I had exams at my university this week
I'll try to fix this and will let you know

guitargeek commented:
No problem! And thank you very much for your work. I'm glad to see some C++ code that calls into Python getting replaced, because this is some of the most fragile code in ROOT that is difficult to maintain :+1: Your implementation in Python makes more sense.

PrasannaKasar commented:
@guitargeek can you please run the CI once?
I have reverted the `CMakeLists.txt` file of the `Pythonization` folder back to its previous version
Also imported numpy locally in the parser functions
And if anything still fails, I have added print statements in the test function to print the versions of Tensorflow/Keras used by the CI

PrasannaKasar commented:
Hi @guitargeek I have few questions regarding the updates you made in the draft pr
<img width="272" height="216" alt="image" src="https://github.com/user-attachments/assets/5a838d81-df1c-4611-9824-cb8d40f5c6f8" />
1. if we don't expose the RModel Keras parser class in C++, how can we run it?
Since the parser is using C++ methods to parse the Keras layers into the rmodel object
2. I really didn't understand the difference between Pythonization of a C++ class and a regular Python class. Could you please help me better understand this difference?

PrasannaKasar commented:
>I don't understand your concern there, can yo clarify? Your parser is a Python class that calls into C++ methods via the ROOT Python interface. The parsers doesn't need to be a C++ class itself for that.

Hi @guitargeek I got it. It was a misunderstanding from my end, I thought we need the parser to be used in C++ as well. Hence, I was insisting on declaring a dummy class in C++. But I'll use the Python class itself now

guitargeek commented:
Yes, correct!

guitargeek commented:
But we should check with @sanjibansg if you want to continue supporting model parsing in C++ via the `RSofieReader`. If yes, this should be implemented using the existing abstraction to call Python code from C++, called [TPython](https://root.cern/doc/master/classTPython.html). Then you have to modify the [RSofieReader](https://github.com/root-project/root/blob/master/tmva/tmva/inc/TMVA/RSofieReader.hxx#L111) to use your new parser class via TPython.

In other words, calling from C++ into Python should only happen via TPython, and not by including `<Pyhon.h>` and using the Python C API directly, as it is happening right now in the old implementation of the Keras parser:
https://github.com/root-project/root/blob/master/tmva/pymva/src/RModelParser_Keras.cxx#L22

PrasannaKasar commented:
@guitargeek I have been getting this issue since the first CI
```
  Default AvgPoolingOp only supports NHWC on device type 
  [[{{node model_9/average_pooling2d/AvgPool}}]] [Op:__inference_train_function_2459]
```

I looked into the Keras documentation but I could find only few mentions of it in their issues section and found out that the issue is not limited to any Keras version, but it appears to be fixed in the later versions of Keras 2
I also tried looking into the logs while creating the jobs, but I could not find any specific file for `Mac14, Mac15, mac 26` (These are some of the docker images where the tests are failing)
I could find the log file for the Fedora though where a similar issue is occurring and found out that the Keras 3.20.0 was being used. I tried training the same `AveragePool` model on my machine (using CPU), but I didn't encounter it.
Can you please help me with this?

PrasannaKasar commented:
@guitargeek I have informed Sanjiban about this issue
In my opinion we should disable the tests which give this `NCHW not supported` error
I'll let you know on this as soon as I get an update

bellenot commented:
> don't these platforms use Keras? [Windows 10 x86 Release](https://github.com/root-project/root/actions/runs/19207660928/job/55016451294#logs) [fedora43](https://github.com/root-project/root/actions/runs/19207660928/job/55016451337#logs)

No, there is no Keras installed on our Windows runners. And as a general comment, one must test if Keras (or any other package) is available before trying to run any test depending on it
